Support environment variables in Cloudflare and Netlify Edge functions (#5301)
withastro/astro: packages/integrations/cloudflare/README.md

## Environment Variables

> As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), private environment variables should only be accessed when a request has happened. Usually, this means moving environment variable access inside a function.

Author: Hmm using a function is not the only way to workaround this issue, so it feels a bit odd to directly say "use a function". You can use `import.meta.env.PRIVATE_VAR` in an `.astro` file without a function for example.

I'm also not sure how to explain this well as a "request" is generally a Cloudflare implementation detail, so I wrote this as "if you see this issue, try this solution". I'm still slightly leaning on the original because of this though. 😅

> As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), private environment variables should only be accessed when a request has happened. Usually, this means moving environment variable access inside a function.

Reviewer: OK, yah, if that's not the only solution, makes sense!

Can you revisit the original re: the word "should" though? That's a tricky word. It can be really difficult to tell whether "private environment variables should only be accessed when a request has happened." means:

You must take care to only access them when..(best practice) 

vs.

They are likely only going to be accessed (probability/what's likely to happen)

It was really that one single word that made me think maybe we could get clearer with a restructuring!

> As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), private environment variables should only be accessed when a request has happened. Usually, this means moving environment variable access inside a function.

Author: Good point, perhaps something like below could be more assertive:

```suggestion
As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), you can only access private environment variables when a request has happened. Usually, this means moving environment variable access inside a function.
```

> As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), private environment variables should only be accessed when a request has happened. Usually, this means moving environment variable access inside a function.

Reviewer: I like it! Ship it! 🚀

```js
// util.js

// ❌ Do not access environment variables on initialization
const serverUrl = import.meta.env.SERVER_URL;

export function fetchData() {
  // ✅ Access inside a function instead

>   return fetch(import.meta.env.SERVER_URL + "/api/data");

Author: I learnt the "no wrong code" rule from y'all too, but I can't quite think of an alternative to show what not to do. Maybe it would be clearer if this shows an endpoint and have `const serverUrl = import.meta.env.SERVER_URL;` inside the function with extra comments. Will tweak this!

>   return fetch(import.meta.env.SERVER_URL + "/api/data");

Author: Added it in [43303e3](https://github.com/withastro/astro/pull/5301/commits/43303e36990d4261bda67b6bc88ed3bbb6cb6214). It doesn't cover for cases where an Astro file imports a `util.js` that does fetch calls though like the previous example, but it feels like the endpoint case would be more common.

## Environment Variables

As Cloudflare Pages Functions [provides environment variables per request](https://developers.cloudflare.com/pages/platform/functions/#adding-environment-variables-locally), private environment variables should only be accessed when a request has happened. Usually, this means moving environment variable access inside a function.

```js
// pages/[id].json.js

export function get({ params }) {

>   // Access environment variables inside a function

Reviewer: ```suggestion
  // Access environment variables per request inside a function
```
Do you think this is helpful? Since we're not showing "Hey don't don't do this anymore!" maybe the extra nugde of "per request" repeated from above will help connect the idea of being in a function? Totally your call.

Otherwise, I'm happy to let you decide whether you think this is better or worse than before. I *do* like that it's more of a complete example, though! So if you think this works, I'd lean towards this now!

Author: It does read a bit more succinct. LGTM!
